docs/langref/relay/type_system.rst

```diff
@@ -0,0 +1,187 @@
+on the expressivity of the input language.
+
+Because Relay is intended as an IR we require *some* type information to provide
+full inference. We don't believe this to be an issue as many of the IR builder
+inferfaces require some type information, or can generate IR based on their own
+higher level inferences.
+
+We view this limited shape inference as a simpler form of type
+inference. Instead of relying on an ad-hoc procedure for recovering type
+information from a potentially dynamic program, we apply ideas from compiler and IR design.
+
+Below we briefly dicsuss the different kinds of types in Relay.
+
+=====
+Types
+=====
+
+Relay's type system has a "language of types" which allow us to write down the type of
+a Relay program. Below we detail the langauge of types and how we assign them to Relay
+programs.
+
+Type
+~~~~
+The base type for all Relay types. All Relay types are sub-classes of this base type.
+
+See :py:class:`~tvm.relay.type.Type` for its definition and documentation.
+
+Tensor Type
+~~~~~~~~~~~
+
+A concrete TensorType in Relay, see tvm/relay/type.h for more details.
+
+This is the type assigned to tensor's with a known dype and shape. For
+example a tensor of `float32` and `(5, 5)`.
+
+
+
+Tensor values in Relay are typed with tensor types. A tensor type is
+parametrized by a data type, and shape. The data type must be a base
+type as enforced by the kind checking rules described in TODO.
+
+This restriction importantly means
+
+The shape may be any valid Relay shape as described in the below
+section on shapes.
+
+
+See :py:class:`~tvm.relay.type.TensorType` for its definition and documentation.
+
+Kind
+~~~~
+The kind of a type parameter, represents a variable shape,
+base type, type, or dimension.
+
+This controls what a type parameter is allowed to be instantiated
+with. For example one's of kind BaseType can only be `float32`,
+`int32`, and so on.
+
+See :py:class:`~tvm.relay.type.Kind` for its definition and documentation.
+
+Type Parameter
+~~~~~~~~~~~~~~
+
+A type parameter used for generic types in Relay,
+see tvm/relay/type.h for more details.
+
+A type parameter represents a type placeholder which will
+be filled in later on. This allows the user to write
+functions which are generic over types.
+
+See :py:class:`~tvm.relay.type.TypeParam` for its definition and documentation.
+
+Type Constriant
+~~~~~~~~~~~~~~~
+
+Abstract class representing a type constraint, to be elaborated
+on in further releases.
+
+See :py:class:`~tvm.relay.type.TypeConstraint` for its definition and documentation.
+
+Function Type
+~~~~~~~~~~~~~
+A function type in Relay, see tvm/relay/type.h for more details.
+
+This is the type assigned to functions in Relay. They consist of
+a list of type parameters which enable the definition of generic
+functions, a set of type constraints which we omit for the time
+being, a sequence of argument types, and a return type.
```

Comment:
"which we omit for the time being" is a parenthetical remark, so it should be offset from the surrounding context with parentheses or dashes